crates/misc/component-fuzz-util/src/lib.rs

```diff
@@ -773,34 +796,42 @@ pub struct TestCase {
 impl TestCase {
     /// Generate a `Declarations` for this `TestCase` which may be used to build a component to execute the case.
     pub fn declarations(&self) -> Declarations {
-        let mut types = String::new();
-        let name_counter = &mut 0;
-
-        let params = self
-            .params
-            .iter()
-            .map(|ty| {
-                let mut tmp = String::new();
-                write_component_type(ty, &mut tmp, name_counter, &mut types);
-                format!("(param {tmp})")
-            })
-            .collect::<Box<[_]>>()
-            .join(" ")
-            .into();
-
-        let result = {
-            let mut tmp = String::new();
-            write_component_type(&self.result, &mut tmp, name_counter, &mut types);
-            format!("(result {tmp})")
+        let mut builder = TypesBuilder {
+            next: 0,
+            worklist: Vec::new(),
+        };
+
+        let mut params = String::new();
+        for ty in self.params.iter() {
+            params.push_str(" (param ");
+            builder.write_ref(ty, &mut params);
+            params.push_str(")");
         }
-        .into();
+
+        let mut result = String::from("(result ");
+        builder.write_ref(&self.result, &mut result);
+        result.push_str(")");
 
         let import_and_export = make_import_and_export(&self.params, &self.result);
 
+        let mut type_decls = Vec::new();
+        while let Some((idx, ty)) = builder.worklist.pop() {
+            type_decls.push(builder.write_decl(idx, ty));
+        }
+
+        // Note that types are printed here in reverse order isnce they were
+        // pushed onto `type_decls` as they were referenced meaning the last one
+        // is the "base" one.
+        let mut types = String::new();
+        for decl in type_decls.iter().rev() {
```

Comment:
I guess you might as well consume the vector and drop the partial strings as you go, right?

```suggestion
        for decl in type_decls.into_iter().rev() {
```